src/auth.rs

```diff
@@ -0,0 +1,247 @@
+//! A wrapper for performing Remote Framebuffer authentication.
+
+use anyhow::{bail, Context, Result};
+
+use futures::{SinkExt, StreamExt};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+/// What kind of authentication to use for an RFB connection.
+pub enum RfbAuthentication {
+    /// A null authentication. It does not perform the initial handshake, so it relies on the rest
+    /// of the connection to pass through the data as-is without parsing.
+    Null,
+
+    /// An authentication that parses the initial ProtocolVersion and Security handshakes, and
+    /// passes them as-is to the peer, without acting on it. This leaves the stream in a state
+    /// where the ClientInit handshake is expected to appear next, followed by a stream of normal
+    /// RFB messages can appear.
+    Passthrough,
+}
+
+/// A way of authenticating an RFB connection.
+pub async fn authenticate<SocketStream, WebSocketStream>(
+    authentication: &RfbAuthentication,
+    stream: &mut SocketStream,
+    ws_stream: &mut WebSocketStream,
+) -> Result<()>
+where
+    SocketStream: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin + Send,
+    WebSocketStream: futures::Sink<
+            tokio_tungstenite::tungstenite::protocol::Message,
+            Error = tokio_tungstenite::tungstenite::error::Error,
+        > + futures::Stream<
+            Item = std::result::Result<
+                tokio_tungstenite::tungstenite::protocol::Message,
+                tokio_tungstenite::tungstenite::error::Error,
+            >,
+        > + Unpin
+        + Send,
+{
+    match authentication {
+        RfbAuthentication::Null => Ok(()),
+        RfbAuthentication::Passthrough => authenticate_passthrough(stream, ws_stream).await,
+    }
+}
+
+async fn authenticate_passthrough<SocketStream, WebSocketStream>(
+    stream: &mut SocketStream,
+    ws_stream: &mut WebSocketStream,
+) -> Result<()>
+where
+    SocketStream: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin + Send,
+    WebSocketStream: futures::Sink<
+            tokio_tungstenite::tungstenite::protocol::Message,
+            Error = tokio_tungstenite::tungstenite::error::Error,
+        > + futures::Stream<
+            Item = std::result::Result<
+                tokio_tungstenite::tungstenite::protocol::Message,
+                tokio_tungstenite::tungstenite::error::Error,
+            >,
+        > + Unpin
+        + Send,
+{
+    let mut buf = [0u8; 1024];
+
+    // ProtocolVersion handshake.
+    let n = stream.read(&mut buf[0..12]).await?;
+    if n != 12 {
+        bail!("unexpected server handshake: {:?}", &buf[..n]);
+    }
+    log::debug!("<-: {:?}", std::str::from_utf8(&buf[0..12])?);
+    ws_stream
+        .send(tokio_tungstenite::tungstenite::protocol::Message::Binary(
```

Comment:
wdyt about adding a use for `Message`?
```
use tokio_tungstenite::tungstenite::protocol::Message;
```
or if `Message` seems to general
```
use tokio_tungstenite::tungstenite::protocol::Message as WsMessage;
```

Reply:
AH the `as WsMessage` was the bit i needed! i avoided adding it because it was conflicting with `prost::Message`, but this takes care of that n_n.